### ISSUE TITLE
File sink healthcheck non-functional

### ISSUE THREAD

**Issue by jacksgt** [vectordotdev/vector#17499]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

Hi,

I configured the ["file" sink](https://vector.dev/docs/reference/configuration/sinks/file/) and enabled [sink healthchecks](https://vector.dev/docs/reference/cli/#vector_require_healthy).
The documentation says:

> Exit on startup if any sinks fail healthchecks.

However, it seems that this healthcheck is not implemented (correctly or at all) for the "file" sink.
When vector starts up, the healthcheck passes, but when vector wants to write the first log entry into the file an error occurs.
(see logs below)

In my opinion, vector should check *at startup* if it can write into the designated file.

### Configuration

```text
data_dir = "/var/lib/vector"

[sources.syslog]
type = "syslog"
address = "0.0.0.0:514"
mode = "udp"

[sinks.local_file]
type = "file"
inputs = [ "syslog" ]
path = "/var/lib/vector/%Y-%m-%d.log.json.zstd"
compression = "zstd"
encoding.codec = "json"
```

### Version

vector 0.29.1 (x86_64-unknown-linux-gnu 74ae15e 2023-04-20 14:50:42.739094536)

### Debug Output

```text
Log output:


vector --config-dir /etc/vector --require-healthy true

2023-05-25T12:17:44.321236Z  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info,buffers=info,lapin=info,kube=info"
2023-05-25T12:17:44.322640Z  INFO vector::app: Loading configs. paths=["/etc/vector"]
2023-05-25T12:17:44.329933Z  INFO vector::topology::running: Running healthchecks.
2023-05-25T12:17:44.329976Z  INFO vector::topology::builder: Healthcheck passed.
2023-05-25T12:17:44.329993Z  INFO vector::topology::builder: Healthcheck passed.
2023-05-25T12:17:44.330009Z  INFO vector::topology::running: All healthchecks passed.
2023-05-25T12:17:44.330158Z  INFO vector: Vector has started. debug="false" version="0.29.1" arch="x86_64" revision="74ae15e 2023-04-20 14:50:42.739094536"
2023-05-25T12:17:44.330387Z  INFO source{component_kind="source" component_id=syslog component_type=syslog component_name=syslog}: vector::sources::syslog: Listening. addr=0.0.0.0:50514 type="udp"
2023-05-26T08:33:22.998628Z ERROR sink{component_kind="sink" component_id=local_file component_type=file component_name=local_file}: vector::internal_events::file: Unable to open the file. path=b"/var/lib/vector-2023-05-26.log.json.zstd" error=Permission denied (os error 13) error_code=failed_opening_file error_type="io_failed" stage="sending" internal_log_rate_limit=true

```
```


### Example Data

_No response_

### Additional Context

_No response_

### References

_No response_

**Comment by jszwedko** [comment]
This could definitely be made clearer in the docs, but the `file` sink doesn't have a health check. I am wondering what would be a reasonable think to check at start-up for that sink, though, since the file path is typically a template so we couldn't, for example, just try to `touch` a file. Is there anything you would expect to be checked?

**Comment by zamazan4ik** [comment]
> Is there anything you would expect to be checked?

Well, we could try to replace all templates with some "defaults" like the current time. Stuff that cannot be replaced easily - try to truncate or something like that and then `touch` the resulting path. I guess IRL the most obvious template would be like "static_path/filename_with_template.txt" so in this case we will `touch` a file in the static directory and if a Vector instance has no corresponding permissions - exit during the startup. However, not sure how important is to have a "proper" healthcheck for the File sink.

**Comment by jacksgt** [comment]
Hi everyone, thanks for your comments!

> but the file sink doesn't have a health check. 

Aha, that's the first interesting data point. :-)

> I am wondering what would be a reasonable think to check at start-up for that sink, though, since the file path is typically a template so we couldn't, for example, just try to touch a file. Is there anything you would expect to be checked?

Like @spencergilbert mentioned, from a sysadmin perspective I expect the filepath of `now()` to be checked (i.e. the startup time). Clearly, this doesn't guarantee that the same template will still work tomorrow (e.g. with directories like `/logs/%Y/%m/%d.log`), but no healthcheck can ensure it will work in the future indefinitely - I just expect a startup healthcheck to give me an immediate feedbackup on startup :-)

**Comment by spencergilbert** [comment]
> Clearly, this doesn't guarantee that the same template will still work tomorrow (e.g. with directories like /logs/%Y/%m/%d.log), but no healthcheck can ensure it will work in the future indefinitely - I just expect a startup healthcheck to give me an immediate feedbackup on startup :-)

Yeah, definitely some granularity concerns - we could get really clever and complicated by adjusting the healthcheck based on the provided template. Think worth starting small and going from there makes the most sense, eg. take the _non_-template part of the path and seeing what permissions we have there, and if they seem sufficient.
